satellite/gc/service.go

```diff
@@ -0,0 +1,127 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gc
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/auth/signing"
+	"storj.io/storj/pkg/overlay"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/uplink/piecestore"
+)
+
+// Config contains configurable values for garbage collection
+type Config struct {
+	Interval time.Duration `help:"how frequently garbage collection filters should be sent to storage nodes" releaseDefault:"168h" devDefault:"10m"`
+	Active   bool          `help:"set if garbage collection is actively running or not" releaseDefault:"true" devDefault:"true"`
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int64   `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log             *zap.Logger
+	config          Config
+	transport       transport.Client
+	overlay         overlay.DB
+	lastSendTime    time.Time
+	lastPieceCounts map[storj.NodeID]int
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, config Config, transport transport.Client, overlay overlay.DB) *Service {
+	return &Service{
+		log:             log,
+		config:          config,
+		transport:       transport,
+		overlay:         overlay,
+		lastPieceCounts: make(map[storj.NodeID]int),
+	}
+}
+
+// NewPieceTracker instantiates a piece tracker
+func (service *Service) NewPieceTracker() PieceTracker {
+	// Creation date of the gc bloom filter - the storage nodes shouldn't delete any piece newer than this.
+	filterCreationDate := time.Now().UTC()
+
+	if !service.config.Active || filterCreationDate.Before(service.lastSendTime.Add(service.config.Interval)) {
```

Comment:
Isn't the result of NewPieceTracker() persisted (and thereby always be useless)?  I feel like this whole noOpPieceTracker struct is a bad idea and that this check should be elsewhere.

Reply:
It's reset here when an iteration finishes. Do you think you could explain more of your worries about noOpPieceTracker? also, where do you think a better place for the check would be? https://github.com/storj/storj/pull/2366/files/b43086f0f79b71c76ed97380594c77894ce29e0a#diff-8762ff585be3a81707f74a689d803d6cR151

Reply:
maybe I just need a better explanation of why its necessary

Reply:
I think the idea is that we don't want pieceTracker to be tracking pieces (creating retainInfos, etc.) when it's not time for garbage collection to run. This ```if !service.config.Active || filterCreationDate.Before(service.lastSendTime.Add(service.config.Interval)) { ``` is checking if gc is due for an iteration. If not, then we use noOpPieceTracker which does nothing when the PieceTracker methods are called. An alternative is maybe adding this time check within each method and just have one pieceTracker implementation 🤷🏻‍♀️

Reply:
I talked to Moby and a more in favor of allowing PieceTracker to be nil instead of using to Noop one, but this is good enough I suppose.

Reply:
I find that having fewer conditionals (i.e. checking in several different places whether `pieceTracker` is nil before using it) leads to code that is easier to reason about and easier to change later. The cost in this case is having multiple things that implement the same interface, which is generally an easy thing to reason about, being enforced by the compiler. I like that tradeoff.